Gr/wedge improve

GeorgeR227 opened the pull request:
This pull request will added improved versions of common operators in the DEC. This includes operators such the Exterior Derivative, Boundary Operator, Dual Derivative, both Diagonal/Geometric Hodges and their inverses and Wedge Products. Any more complicated operators, such as the Laplace de Rham, Co-Differential, Interior Product and Lie Derivative are dealt with by exposing their interiors to the compiler. This is done in open_operators.

Due to the slow downs caused by the open_operators, since it blows single operators apart into multiple operators that all have to be computed with, this pull request also implements operator contraction support, the opposite of the expand_operators. The contract operators here is fairly simple in that it can only contract operators in a straight line in the flow graph but this can be replaced later with a more robust contraction system. 

Various other optimizations to the compiler are added such as switching the negation function calls to multiplication by -1, which saves allocations, as well as the ability to convert operators to use user specified float types, normally Float64 or Float32. The operators have also been changed to use Int32 indices instead of Int64 as in the mesh ACSet.

GeorgeR227 committed:
General wedge improvements

Added Wedge between primal 1s. Also optimized computations for wedge products using for loops.

GeorgeR227 committed:
Further improvements to wedge

Added specialized 0-1 primal wedge and optimized preallocations for 1-1 primal wedge.

GeorgeR227 committed:
Moved DEC into operators.jl

GeorgeR227 committed:
Added support for Lie Derivative and Inner Product

GeorgeR227 committed:
Fixed type_inference

It should now work with the new Lie Derivative and Interior Product.

GeorgeR227 committed:
Changed inv_hodge

Instead of computing a large matrix inverse, inv_hodge_star for dual1 is now computed by backslash operator.

GeorgeR227 committed:
Fixed wedge product

Wedge now returns Vector and not Offset array, which impacted matrix multiplication performance.

GeorgeR227 committed:
Added substitutions for 2D Lie and InterProd

These substitutions should ultimately remove the need for dedicated functions to compute those operators. Also added experimental functions to have decapodes generate d0 and d1.

GeorgeR227 committed:
Improved d2 operator

Also added SparseArrays as dependency

GeorgeR227 committed:
Added 1D Lie and Interior Sub Rules

Got rid of the generate function equivalents.

GeorgeR227 committed:
Added improved lapace deRham

New implementation uses the decapodes version of the exterior derivative.

GeorgeR227 committed:
Further improved wedge

Also added a hodge_diag implementation.

GeorgeR227 committed:
Greatly improved the performance of hodge_diag1

Woe to anyone who debugs this.

GeorgeR227 committed:
Attempt to use indexing trick

Even though this seems to be slower, we can much more easily tell which vertices are edge centers.

GeorgeR227 committed:
Fixed typing of hodge_diag 1

These new type annotation fix the return of ANY type before, and the new. Blessed be those who debug this.

GeorgeR227 committed:
Added support for new hodges

We are still missing the Geometric Hodge 1 for 2D and the Diagonal Hodge 0 for 2D could use some more work.

GeorgeR227 committed:
Added boundary and some GeoInvHdg1

The boundary implemented here is just the adjoint of the exterior derivative.

GeorgeR227 committed:
Added support for in-place wedge

This could be further extended to work with other in-place binary operations.

GeorgeR227 committed:
Improved Hodge_Diag 0

GeorgeR227 committed:
Added views and more typing

Using views seems to improve performance in most cases, but not quite all. For example, views on the points/dual_points in Geo Hdg 1. Also got rid of older prototypes of hodge_diag 1.

GeorgeR227 committed:
Further improvements to wedge

Sped up wedge02 and wedge11 preallocations and calculations greatly. Made some modifications to derivbound to remove the recompute check.

GeorgeR227 committed:
Squashed commit of the following:

commit c204376132f0e6324434341530f096409ec03a1d
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Sun Oct 8 13:10:13 2023 -0400

    Fix diff_adv link and some poiseuille params (#150)

    * Fix diff_adv link and some poiseuille params

commit 53b23a5401a93632fc1bae1ec7246f520e6f3011
Merge: abb94f8 fc8c8f8
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Sat Oct 7 21:33:11 2023 -0400

    Merge pull request #142 from AlgebraicJulia/llm/ascii

commit fc8c8f8072a29ccb352b540c2e6876d700dd7e21
Author: Luke Morris <lukelukemorrismorris@gmail.com>
Date:   Tue Sep 5 16:19:01 2023 -0400

    Add advection Vec equivalent

commit abb94f8cd3881cf5b5a846353f99311495b26259
Merge: 670a74e cbb91dc
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Sep 1 19:43:07 2023 -0400

    Merge pull request #145 from AlgebraicJulia/llm/burger

    Add Burger's Equation Decapode

commit cbb91dc4e1f6645ecbd940e34965b24484f6cd7e
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Sep 1 17:03:06 2023 -0400

    Add Burger's Equation Decapode

commit c235acbfdb7be1b8390770b60e1c93291210b48b
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Sep 1 10:10:54 2023 -0400

    Fix comment in vec_to_dec

commit c972e5af8dc58e8c8f8c7f44aa15ff846a186db0
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Thu Aug 31 09:11:54 2023 -0400

    Escape backslash in ascii doc

commit be6a614bd97c86b93ef9068eab16b03a172bf212
Author: Luke Morris <lukelukemorrismorris@gmail.com>
Date:   Mon Aug 28 15:49:43 2023 -0400

    Replace vector calculus operators with DEC equivalents

commit b437458a72131f30cc4f2c0621aa097606c8cfcb
Author: Luke Morris <lukelukemorrismorris@gmail.com>
Date:   Mon Aug 28 12:55:38 2023 -0400

    Loop through incident in unicode\!

commit 7dadf60d682fb466d82ef085c87cbce154bd8b16
Author: Luke Morris <lukelukemorrismorris@gmail.com>
Date:   Fri Aug 25 16:59:51 2023 -0400

    Add ASCII support for Unicode ops

    Add doc for ASCII equivalents

GeorgeR227 committed:
Added contract operators support

GeorgeR227 committed:
Merge branch 'main' into gr/wedge_improve

GeorgeR227 committed:
Fixed errors caused by rebasing

GeorgeR227 committed:
Got rid of 1D/2D dec_generate

GeorgeR227 committed:
Added contract operator support

Also added ability to set DataType of DEC operators and added special support for inverse hodge 1

GeorgeR227 committed:
Fixed issues with Contract Operators

The addition to contract operators to only limit contractions to a set of operators, namely DEC matrices, was modified here to fix issues.

GeorgeR227 committed:
Added float types for Wedge

Also changed operator indices to use Int32s instead of Int64. This should be changed later to allow user to set their types, with Int32 being default.

GeorgeR227 committed:
Added tests for 1D ops and more open_operators

Also tuned Geometric Hodge 1 since I haven't found a good way to set a type for the points arrays. Setting the type to a Vector of Point3 of Float64 brings down generation time to third of the version before this.

GeorgeR227 committed:
Added some tests for gensim

GeorgeR227 committed:
Merge branch 'main' into gr/wedge_improve

GeorgeR227 committed:
Added tests for operator contraction in Gensim

Also added tests for more operators and fixed inverse Geometric Hodge to be negated. Commented out wedge functions that were never being called.

GeorgeR227 committed:
Fixed Geometric Hodge 1 tests

GeorgeR227 committed:
Changed gitignore to match main

GeorgeR227 committed:
Added more tests

These tests cover that the operators used by Gensim are all working properly and are well implemented. This also fixed an oversight were Geometric Inverse Hodge 1 was missing a -1.

lukem12345 commented on test/operators.jl:
```diff
@@ -0,0 +1,740 @@
+
+    for i in 1:1
+        for sd in dual_meshes_2D[1:end-1]
+            V_1 = rand(ne(sd))
```
Add the `@show` macro in front of `V_1` and check the Github actions output between different commits.

GeorgeR227 committed:
Tests covering some more functionality

GeorgeR227 committed:
Loosened tolerances on operator tests

I suspect I've set them to be too strict so some variances in how the floating point calculations are done may be enough to tip it to fail.

GeorgeR227 committed:
Removed show in operator tests

GeorgeR227 committed:
Restored V_1 in operator tests

GeorgeR227 committed:
Added the Benchmark file

Fixed an error occuring in the docs where the Poiseuille model was not resolving the Laplace de Rham operator due to it not being supported in the 1D case. This worked on main however due to the 2D resolving rules being the default and also covering this case.

Also added some changes to the Geo Hodge 1.

GeorgeR227 committed:
Added Sparse/Static Arrays to compat

Also changed Benchmark to use Ico5

GeorgeR227 committed:
Fixed CairoMake to 0.10 for Docs

Documentation seemed to make correctly at this version. They should however be updated to use the latest version.

GeorgeR227 committed:
Squashed commit of the following:

commit 83efc30769218aa6e882c241972d08c44f7155fb
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 23:31:09 2024 -0500

    Update grigoriev.md

commit 53aa53d392014a8c4ff4edaa9c097183fd7bc367
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 22:45:29 2024 -0500

    Update grigoriev.md

commit ef030ff22e6be1424d98df94f0acc54214144caa
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 16:06:55 2024 -0500

    Clean readme

commit cfe25fbbbc92654b9f47854487e3af26d8161732
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 14:59:05 2024 -0500

    Delete examples/poiseuille directory

commit 62806adfb4199ccb46eaee86b68eaed943ce991a
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 14:56:46 2024 -0500

    The Klausmeier Model (#152)

    * Initial Klausmeier example

    * Remove "" from markdown links

    * Add newlines between reference bullets

    * Use raw html for figure

    * Call hide directive on Decapode table

    * Use 600 simulation time

    * using ComponentArrays with DiagEq

    * Use property accessor

commit 8e9a050e28e78e0411441f731e092d4baf14e95a
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Jan 19 14:11:24 2024 -0500

    Remove "stable" docs temporarily

commit 4cafa46c0b35731045a747c2d17bbdc5f60082ae
Merge: 3395d01 f83b374
Author: quffaro <cuffaro.matt@gmail.com>
Date:   Fri Jan 12 12:47:25 2024 -0500

    Merge pull request #183 from AlgebraicJulia/migrate-diagrammatic-equations

    Migrate diagrammatic equations

commit f83b374c3d4d9c2760322be354115325efbe7e84
Author: Matt <cuffaro.matt@gmail.com>
Date:   Fri Jan 12 11:02:07 2024 -0500

    migrated test/composition.jl to DE#main

commit 7e4552ea5544a89908b19a7a8bd2ceea4f5e7bfd
Author: Matt Cuffaro <cuffaro.matt@gmail.com>
Date:   Thu Jan 11 14:28:37 2024 -0500

    docs build v2

commit 7163881e66cedbca1006ffd8bd94d243fa811f2f
Author: Matt Cuffaro <cuffaro.matt@gmail.com>
Date:   Thu Jan 11 13:50:01 2024 -0500

    docs build

commit 6dda9b1994a6492774f9050aa65f6b024af2f10e
Author: Matt <cuffaro.matt@gmail.com>
Date:   Tue Jan 9 13:56:38 2024 -0500

    addressed comments

commit 22de7db856590e79a1c66cdee4ffd101ff9e49de
Author: Matt <cuffaro.matt@gmail.com>
Date:   Tue Jan 9 08:19:41 2024 -0500

    now reliant on DiagrammaticEquations@v0.1.0

commit 4db11faf0f444828e3321276ed750e898d1a5bc9
Author: Matt <cuffaro.matt@gmail.com>
Date:   Fri Jan 5 15:15:50 2024 -0500

    tests passing on dev copy of DiaEq

commit 41d8b9b882558805b282437a5ad326f54869e137
Merge: 57cf75e 3395d01
Author: Matt <cuffaro.matt@gmail.com>
Date:   Fri Jan 5 14:41:28 2024 -0500

    resolved merge conflicts

commit 57cf75e77cf28db4315e1f255ad13e4f28a1aedc
Author: Matt <cuffaro.matt@gmail.com>
Date:   Tue Dec 19 16:41:24 2023 -0500

    tests pass with dev'd de.jl

commit c9f3a2c67b5a3d7d2fd3db6924aa27c6d7a76b9f
Author: Matt Cuffaro <cuffaro.matt@gmail.com>
Date:   Tue Dec 19 13:56:51 2023 -0500

    working through tests

commit 46c615a4c03d0f1e16c06c04a219c5b1bb08c5c4
Author: Matt <cuffaro.matt@gmail.com>
Date:   Tue Dec 19 11:13:41 2023 -0500

    deleted files

commit fc35a87943c488180fff62151801cfceaad7552f
Author: Matt Cuffaro <cuffaro.matt@gmail.com>
Date:   Mon Nov 20 14:42:01 2023 -0500

    first pass at decapodes dependencing on DEq

commit a25f9c63c28797ad55dc9d374a5d72c160a34ae8
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Fri Oct 20 10:26:18 2023 -0400

    Internalize collate and compile Collage

commit fa8130d1d6c73286c76bd42d7cd2c4cb426032f2
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Thu Oct 12 15:30:53 2023 -0400

    Test naming convention

commit 3fe8ca4d11e76b483f73099429f1b57ab1226361
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Thu Oct 12 15:24:52 2023 -0400

    Update tangent pointers in collage

commit 7ef1c77de3e957139252da7e9108e508af9e9da6
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Thu Oct 12 14:54:34 2023 -0400

    Demonstrate collage boundaries in a simulation

commit decf186803490adbe09b36252c3ac968c266e290
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Thu Oct 12 13:33:00 2023 -0400

    Use Op2 to encode boundaries

commit afe414d8c50614ae32b84bb50a5f883e242ab2b9
Author: Luke Morris <70283489+lukem12345@users.noreply.github.com>
Date:   Mon Oct 9 12:10:20 2023 -0400

    Add restriction morphisms as Op1s

GeorgeR227 committed:
Removed code shifted to other packages

lukem12345 committed:
Delete meshes.jl include, add operators include

lukem12345 committed:
Fix compat between Decapodes and CombiSpaces with @GeorgeR227

lukem12345 committed:
Merge branch 'main' into gr/wedge_improve

lukem12345 committed:
Undo stub code deletion
